src/group_theory/quaternion_group.lean

```diff
@@ -0,0 +1,289 @@
+  right_inv := by rintro (x | x); refl }
+
+/-- The special case that more or less by definition `quaternion_group 0` is isomorphic to the
+infinite dihedral group. -/
+def quaternion_group_zero_equiv_dihedral_zero : quaternion_group 0 ≃* dihedral 0 :=
+{ to_fun :=  λ i, match i with
+                | (a j) := dihedral.r j
+                | (xa j) := dihedral.sr j
+                end,
+  inv_fun := λ i, match i with
+                | (dihedral.r j) := a j
+                | (dihedral.sr j) := xa j
+                end,
+  left_inv := by rintro (k | k); refl,
+  right_inv := by rintro (k | k); refl,
+  map_mul' := begin rintros (k | k) (l | l); try {refl},
+                    change quaternion_group_zero_equiv_dihedral_zero._match_1 (a (0 + l - k)) = _,
+                    rw zero_add,
+                    refl end }
+
+/-- Some of the lemmas on `zmod m` require that `m` is positive, as `m = 2 * n` is the case relevant
+in this file but we don't want to write `[fact (0 < 2 * n)]` we make this lemma a local instance. -/
+private lemma two_mul_pos_of_pos [fact (0 < n)] : fact (0 < 2 * n) :=
+begin
+rw ← zero_mul 0,
+          apply mul_lt_mul',
+            { exact zero_le 2 },
+            { assumption },
+            { exact zero_le 0 },
+            { exact nat.zero_lt_succ 1 }
+end
+
+local attribute [instance] two_mul_pos_of_pos
+
+/--
+If `0 < n`, then `quaternion_group n` is a finite group.
+-/
+instance [fact (0 < n)] : fintype (quaternion_group n) := fintype.of_equiv _ fintype_helper
+
+instance : nontrivial (quaternion_group n) := ⟨⟨a 0, xa 0, dec_trivial⟩⟩
+
+/--
+If `0 < n`, then `quaternion_group n` has `4n` elements.
+-/
+lemma card [fact (0 < n)] : fintype.card (quaternion_group n) = 4 * n :=
+begin
+  rw ← fintype.card_eq.mpr ⟨fintype_helper⟩,
+  change fintype.card (zmod (2 * n) ⊕ zmod (2 * n)) = 4 * n,
+  rw [fintype.card_sum, zmod.card, two_mul],
+  ring
+end
+
+@[simp] lemma a_one_pow (k : ℕ) : (a 1 : quaternion_group n) ^ k = a k :=
+begin
+  induction k with k IH,
+  { refl },
+  { rw [pow_succ, IH, a_mul_a],
+    congr' 1,
+    norm_cast,
+    rw nat.one_add }
+end
+
+@[simp] lemma a_one_pow_n : (a 1 : quaternion_group n)^(2 * n) = 1 :=
+begin
+  cases n,
+  { simp_rw [mul_zero, pow_zero] },
+  { rw [a_one_pow, one_def],
+    congr' 1,
+    exact zmod.nat_cast_self _ }
+end
+
+@[simp] lemma xa_pow_two (i : zmod (2 * n)) : xa i ^ 2 = a n :=
+begin
+  simp [pow_two]
+end
+
+@[simp] lemma xa_pow_four (i : zmod (2 * n)) : xa i ^ 4 = 1 :=
+begin
+  simp only [pow_succ, pow_two, xa_mul_xa, xa_mul_a, add_sub_cancel, add_sub_assoc, add_sub_cancel',
+             sub_self, add_zero],
+  norm_cast,
+  rw ← two_mul,
+  simp [one_def],
+end
+
+/--
+If `0 < n`, then `xa i` has order 4.
+-/
+@[simp] lemma order_of_xa [fact (0 < n)] (i : zmod (2 * n)) : order_of (xa i) = 4 :=
+begin
+  have h : order_of (xa i) ≤ 4 := order_of_le_of_pow_eq_one (by norm_num) (xa_pow_four i),
+  interval_cases (order_of (xa i)) with hcas,
```

Comment:
I wonder if
```
  change _ = 2^2,
  apply nat.eq_prime_pow_of_dvd_least_prime_pow nat.prime_two,
  { rw order_of_dvd_iff_pow_eq_one,
    intro h,
    sorry },
  { rw order_of_dvd_iff_pow_eq_one, 
    sorry, },
```
would be a better start. Even better would be a lemma that combined `eq_prime_pow_of_dvd_least_prime_pow` and `order_of_dvd_iff_pow_eq_one` so you didn't need the rewrites.

Reply:
It certainly is. I'll work on the combination lemma next.

Reply:
This turned out really nicely, I think. Thanks a lot for the suggestion.